gitlabform/processors/project/project_processor.py

```diff
@@ -1,17 +1,74 @@
-        if configuration["project"].get("archive"):
-            verbose("Archiving project...")
-            self.gitlab.archive(project_and_group)
-        else:
-            verbose("Unarchiving project...")
-            self.gitlab.unarchive(project_and_group)
+        if configuration["project"].get("transfer_from") is not None:
+            source_project_path_with_namespace = configuration["project"].get(
+                "transfer_from"
+            )
+            destination_project_path_with_namespace = project_and_group
+            gl: Gitlab = GitlabWrapper(self.gitlab)._gitlab
+
+            # Check if the project was already transfered (i.e. in previous run) or a project with same path already exists
+            try:
+                project: Project = gl.projects.get(
+                    destination_project_path_with_namespace
+                )
+                verbose(
+                    "Project already exists: '"
+                    + project.path_with_namespace
+                    + "'. Ignoring 'transfer_from' config..."
+                )
+            except GitlabGetError:
+                # Project doesn't exist at the destination. Let's process the transfer request
+                project_to_be_transferred: Project = gl.projects.get(
+                    source_project_path_with_namespace
+                )
+                destination_project_path = project_and_group.split("/")[-1]
+                # Check if the project path needs to be updated; In Gitlab, path maybe different than name
+                if destination_project_path != project_to_be_transferred.path:
+                    verbose(
+                        "Updating the source project path from '"
+                        + project_to_be_transferred.path
+                        + "' to '"
+                        + destination_project_path
+                    )
+                    gl.projects.update(
+                        project_to_be_transferred.id, {"path": destination_project_path}
+                    )
+
+                try:
+                    project_transfer_destination_group, _ = project_and_group.rsplit(
+                        "/", 1
+                    )
+                    verbose(
+                        "Transferring project to '"
+                        + project_transfer_destination_group
+                        + "' group..."
+                    )
+                    project_to_be_transferred.transfer(
+                        project_transfer_destination_group
+                    )
+                except GitlabTransferProjectError:
+                    fatal(
+                        "Encountered error transferring project. Please check if project transfer requirements were met. Docs: https://docs.gitlab.com/ee/user/project/settings/index.html#transfer-a-project-to-another-namespace"
+                    )
+                    raise
```

Comment:
Thanks for the review @gdubicki . I was taking a final glance before merging and a thought occurred.

Should this be a `fatal` error? Will gitlabform continue processing other projects or will it exit here?

Reply:
Right, you should not re-raise it.

I forgot a little how it worked but I refreshed my memory. We have 3 kinds of errors in gitlabform:

1. Regular exceptions - these will break gitlabform run if `--terminate` flag is set in cli, or will just break processing given group or project if not - see [this code](https://github.com/gitlabform/gitlabform/blob/ignore_usernames_case/gitlabform/__init__.py#L397-L427),
2. Unconditionally fatal errors - if `fatal()` is called without an `if`, for example [here](https://github.com/gitlabform/gitlabform/blob/ignore_usernames_case/gitlabform/processors/multiple_entities_processor.py#L190-L194),
3. Conditionally fatal errors - if `--strict` flag is set in cli, if will break a gitlabform run, otherwise it will not. It's restricted to some special cases of given entity, usually a branch, not existing. See [this code](https://github.com/gitlabform/gitlabform/blob/ignore_usernames_case/gitlabform/processors/project/files_processor.py#L49-L56) as an example.

It may seem more chaotic than it actually is.

There is value in having a separate flag for NOT allowing some entities to not exist (`--strict`) and allowing that by default, because in one of the main use cases of this app I used to manage tens of repositories in a single group where some had only `master` branch, some had `develop` and `master` and I could have a single config that did some things with both branch but not fail if `develop` did not exist.

But probably we could make it more consistent by checking all the error cases and maybe expanding the cases that `--strict` covers. 🤔

Reply:
Thanks @gdubicki for the details. I'll look into it and update my exception handling.

Reply:
Hi @gdubicki - I think in this case it makes sense for this to be handled based on `--tarminate` flag. What do you think? I've never had it enabled in my actual usage though. And to be honest, I never quite understood the `--strict` flag (i.e. why another flag in addition to `--terminate`)?. Anyways, I think that's a discussion for separate issue or possibly for next major version design.

Unfortunately I couldn't figure out how to check for the `--terminate` flag in the above exception handling. From the `ProjectProcessor`, I'm not sure how to access it. So, I've just set it as a "warning" message for now. Could you please help when you get a chance? Thank you!

Reply:
@gdubicki - Would you be able to help me with this update?

Reply:
After consideration I think you should not catch the `GitlabTransferProjectError` exception at all. If you won't do that, then [this code](https://github.com/gitlabform/gitlabform/blob/main/gitlabform/__init__.py#L475-L489) with catch it and either break the run (if `--terminate` flag is set) or just break processing given project.

Reply:
Thanks for looking into it @gdubicki . In that case, wouldn't the original setup have worked because I had re-raised the exception after logging a message? Should I revert to that code or do you think simply remove the exception catching/handling?

Reply:
I agree with you. I'll push an update and remove the exception catching so that we're consistent.

We can open a separate issue with a proposal for improving the logging of exception handling in similar cases.

Reply:
Opened a separate issue to keep track of this: https://github.com/gitlabform/gitlabform/issues/611

I've removed the `try` and `except` that was catching the exception earlier. Well... I've commented them out as "Todo". I also realized that the acceptance for it wasn't capturing the custom message. It would pass regardless of what I wrote. Instead of removing the test case, I marked it to be skipped for now. It should probably be enabled and fixed later when exception catching/handling is done with custom messages as per above issue/idea.

Please review one more time.